<a href="https://colab.research.google.com/github/yeialgov/MAYC_ML_Revenue_Prediction/blob/EDA/Yeisson_Copy_of_MAYC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Importing kaggle data 

In [0]:
#Mount the Google Drive to Google Colab
#1. get the library
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


In [0]:
root_path = '/content/gdrive/My Drive/ReDi_DA_Project'

In [7]:
cd '/content/gdrive/My Drive/ReDi_DA_Project'

/content/gdrive/My Drive/ReDi_DA_Project


In [8]:
#for checking the Data
ls -alFh

total 98M
-r-------- 1 root root 98M May  4 09:14 preprocessed_ga_revenue_data.zip


In [9]:
#! unzip train.csv.zip
! unzip preprocessed_ga_revenue_data.zip

Archive:  preprocessed_ga_revenue_data.zip
   creating: preprocessed_ga_revenue_data/
  inflating: preprocessed_ga_revenue_data/train_flat_no_hits.pkl  
  inflating: __MACOSX/preprocessed_ga_revenue_data/._train_flat_no_hits.pkl  
  inflating: preprocessed_ga_revenue_data/test_flat_no_hits.pkl  
  inflating: __MACOSX/preprocessed_ga_revenue_data/._test_flat_no_hits.pkl  
  inflating: preprocessed_ga_revenue_data/train_hits_flat.pkl  
  inflating: __MACOSX/preprocessed_ga_revenue_data/._train_hits_flat.pkl  
  inflating: preprocessed_ga_revenue_data/test_hits_flat.pkl  
  inflating: __MACOSX/preprocessed_ga_revenue_data/._test_hits_flat.pkl  


In [10]:
ls

__MACOSX/  preprocessed_ga_revenue_data/  preprocessed_ga_revenue_data.zip


In [11]:
cd preprocessed_ga_revenue_data

/content/gdrive/My Drive/ReDi_DA_Project/preprocessed_ga_revenue_data


In [12]:
ls

test_flat_no_hits.pkl  train_flat_no_hits.pkl
test_hits_flat.pkl     train_hits_flat.pkl


# Creating dfs

In [0]:
# download the libraries
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
#!pip3 install --upgrade seaborn cufflinks plotly
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
%matplotlib inline
#?
import requests

In [0]:
# load training data set info a df
#df_train=pd.read_csv(filepath_or_buffer='/content/gdrive/My Drive/DA-Data/Kaggel_test.csv')
df_train = pd.read_pickle('train_flat_no_hits.pkl')
#df_test = pd.read_pickle('test_flat_no_hits.pkl')

In [15]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1708337 entries, 0 to 1708336
Data columns (total 60 columns):
 #   Column                                             Dtype 
---  ------                                             ----- 
 0   channelGrouping                                    object
 1   date                                               int64 
 2   fullVisitorId                                      object
 3   socialEngagementType                               object
 4   visitId                                            int64 
 5   visitNumber                                        int64 
 6   visitStartTime                                     int64 
 7   device.browser                                     object
 8   device.browserVersion                              object
 9   device.browserSize                                 object
 10  device.operatingSystem                             object
 11  device.operatingSystemVersion                      object
 12  

In [16]:
df_train.head(5)

channelGrouping      date  ... customDimension.index customDimension.value
0  Organic Search  20171016  ...                     4                  EMEA
1        Referral  20171016  ...                     4         North America
2          Direct  20171016  ...                     4         North America
3  Organic Search  20171016  ...                     4                  EMEA
4  Organic Search  20171016  ...                     4       Central America

[5 rows x 60 columns]

In [0]:
df_train['totals.transactionRevenue'].value_counts()

16990000      308
19990000      248
39980000      220
18990000      219
21990000      211
             ... 
1199840000      1
20290000        1
333340000       1
827100000       1
300060000       1
Name: totals.transactionRevenue, Length: 7251, dtype: int64

In [17]:
df_train.channelGrouping.unique()

array(['Organic Search', 'Referral', 'Direct', 'Paid Search', 'Display',
       'Affiliates', 'Social', '(Other)'], dtype=object)

**fullVisitorId**- A unique identifier for each user of the Google Merchandise Store.
**channelGrouping **- The channel via which the user came to the Store.


**geoNetwork** - This section contains information about the geography of the user.
**socialEngagementType** - Engagement type, either "Socially Engaged" or "Not Socially Engaged".


visitId - An identifier for this session. This is part of the value usually stored as the _utmb cookie. This is only unique to the user. *For a completely unique ID, you should use a combination of fullVisitorId and visitId*.
visitNumber - The session number for this user. If this is the first session, then this is set to 1.
visitStartTime - The timestamp (expressed as POSIX time).


totals - This set of columns mostly includes high-level aggregate data.

In [0]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1708337 entries, 0 to 1708336
Data columns (total 60 columns):
 #   Column                                             Dtype 
---  ------                                             ----- 
 0   channelGrouping                                    object
 1   date                                               int64 
 2   fullVisitorId                                      object
 3   socialEngagementType                               object
 4   visitId                                            int64 
 5   visitNumber                                        int64 
 6   visitStartTime                                     int64 
 7   device.browser                                     object
 8   device.browserVersion                              object
 9   device.browserSize                                 object
 10  device.operatingSystem                             object
 11  device.operatingSystemVersion                      object
 12  

# Question:
not available in demo dataset?

In [25]:
#**device **- The specifications for the device used to access the Store.
df_train['device.browser'][:10].unique()

array(['Firefox', 'Chrome', 'Safari'], dtype=object)

# Question:
meaning of each?

In [28]:
#**totals** - This section contains aggregate values across the session.
df_train.totals[:10].unique()

SyntaxError: ignored

In [22]:
df_train.geoNetwork[:10].unique()

AttributeError: ignored

In [29]:
df_train.socialEngagementType.unique()

array(['Not Socially Engaged'], dtype=object)

# Qusetion:
How can I use it?

In [0]:
#trafficSource - This section contains information about the Traffic Source from which the session originated.
df_train.trafficSource[:10].unique()


array(['{"campaign": "(not set)", "source": "google", "medium": "organic", "keyword": "(not provided)", "adwordsClickInfo": {"criteriaParameters": "not available in demo dataset"}, "isTrueDirect": true}',
       '{"campaign": "(not set)", "source": "google", "medium": "organic", "keyword": "(not provided)", "adwordsClickInfo": {"criteriaParameters": "not available in demo dataset"}}'],
      dtype=object)

In [0]:
#hits - This row and nested fields are populated for any and all types of hits. Provides a record of all page visits.
df_train.hits[:10].unique()

In [0]:
#customDimensions - This section contains any user-level or session-level custom dimensions that are set for a session. This is a repeated field and has an entry for each dimension that is set.


# Question:

In [23]:
#**date** - The date on which the user visited the Store.
date_aggregated.describe()

NameError: ignored

In [0]:
#which are the popular channels the user came to the Store with?

# What the columns mean

Each row in the dataset is one visit to the store. 
We are predicting the log of the total revenue per user.
Be aware that not all rows in test.csv will correspond to a row in the submission, but all unique fullVisitorIds will correspond to a row in the submission.

# Separating JSON written columns

In [0]:
TODO rewrite it 
# j_columns = ['device', 'geoNetwork', 'totals', 'trafficSource'] # Columns that have json format
# #Code to transform the json format columns in table
# def json_to_load(data_Ml):
#     data_Ml = pd.read_csv('trainML.csv',converters={column: json.loads for column in j_columns})
#     for column in j_columns:
#         df_j_columns = json_normalize(data_Ml[column])
#         # here will be set the name using the category and subcategory of json columns
#         df_j_columns.columns = [f"{column}.{subcolumn}" for subcolumn in df_j_columns.columns]
#         #let drop the original columns
#         data_Ml = data_Ml.drop(column, axis=1).merge(df_j_columns,right_index=True, left_index=True)
#     return data_Ml

# Yeisson EDA of geograpfical info

In [30]:
df_train.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1708337 entries, 0 to 1708336
Data columns (total 60 columns):
 #   Column                                             Dtype 
---  ------                                             ----- 
 0   channelGrouping                                    object
 1   date                                               int64 
 2   fullVisitorId                                      object
 3   socialEngagementType                               object
 4   visitId                                            int64 
 5   visitNumber                                        int64 
 6   visitStartTime                                     int64 
 7   device.browser                                     object
 8   device.browserVersion                              object
 9   device.browserSize                                 object
 10  device.operatingSystem                             object
 11  device.operatingSystemVersion                      object
 12  

In [79]:
df_train.agg(['count', 'size', 'nunique'])


channelGrouping     date  ...  customDimension.index  customDimension.value
count            1708337  1708337  ...                1375102                1375102
size             1708337  1708337  ...                1708337                1708337
nunique                8      638  ...                      1                      5

[3 rows x 60 columns]

In [0]:
Geo_columns=['geoNetwork.continent',	'geoNetwork.subContinent',	'geoNetwork.country',	'geoNetwork.region',	'geoNetwork.metro',	'geoNetwork.city',	'geoNetwork.cityId'	,'geoNetwork.networkDomain',	'geoNetwork.latitude',	'geoNetwork.longitude','geoNetwork.networkLocation']

In [81]:
df_train[Geo_columns].agg(['count', 'size', 'nunique'])

geoNetwork.continent  ...  geoNetwork.networkLocation
count                 1708337  ...                     1708337
size                  1708337  ...                     1708337
nunique                     6  ...                           1

[3 rows x 11 columns]

In [0]:
#update the Geo_column list without the not available values
Geo_columns=['geoNetwork.continent',	'geoNetwork.subContinent',	'geoNetwork.country',	'geoNetwork.region',	'geoNetwork.metro',	'geoNetwork.city',	'geoNetwork.networkDomain']

In [86]:
df_train[Geo_columns].agg(['count', 'size', 'nunique'])

geoNetwork.continent  ...  geoNetwork.networkDomain
count                 1708337  ...                   1708337
size                  1708337  ...                   1708337
nunique                     6  ...                     41982

[3 rows x 7 columns]

In [50]:
df_train.groupby(['geoNetwork.continent'])[['totals.transactionRevenue',]].count()

geoNetwork.continent
(not set)        7
Africa          12
Americas     18150
Asia           190
Europe         126
Oceania         29
Name: totals.transactionRevenue, dtype: int64

In [91]:
A=df_train.groupby(['geoNetwork.networkDomain'])[['totals.transactionRevenue']].count()
A=A[A['totals.transactionRevenue']>0]
A.sort_values(by='totals.transactionRevenue', ascending=False, inplace=True)
A

totals.transactionRevenue
geoNetwork.networkDomain                           
(not set)                                     10867
comcast.net                                    1505
verizon.net                                     755
unknown.unknown                                 665
rr.com                                          588
...                                             ...
mtco.com                                          1
mtnsat.com                                        1
mtroyal.ca                                        1
mtsallstream.net                                  1
pmgroup-global.com                                1

[996 rows x 1 columns]

In [92]:
B=df_train.groupby(['geoNetwork.networkDomain'])[['totals.transactionRevenue']].sum()
B=B[B['totals.transactionRevenue']>0]
B.sort_values(by='totals.transactionRevenue', ascending=False, inplace=True)
B

totals.transactionRevenue
geoNetwork.networkDomain                           
(not set)                              1.305843e+12
comcastbusiness.net                    2.079817e+11
comcast.net                            1.773147e+11
verizon.net                            7.967340e+10
unknown.unknown                        5.825398e+10
...                                             ...
netfirms.com                           1.990000e+06
hsn.net                                1.990000e+06
dvois.com                              1.500000e+06
fsu.edu                                1.500000e+06
16clouds.com                           1.500000e+06

[996 rows x 1 columns]

## Domain Insight:
- From the 41982 unique domains register only 996 has had at least one transaction.
- the not set domain aggregate a big number of finished transactions

In [61]:
#transform the objective variable into a numerical variable
df_train["totals.transactionRevenue"] = df_train["totals.transactionRevenue"].astype('float')
df_train.groupby(['geoNetwork.continent'])[['totals.transactionRevenue',]].sum()

totals.transactionRevenue
geoNetwork.continent                           
(not set)                          7.952200e+08
Africa                             9.025990e+09
Americas                           2.265098e+12
Asia                               2.339772e+10
Europe                             1.124745e+10
Oceania                            6.781100e+09